File: go/tests/unit/foundation_decorator_test.go
package unit_test

import (
	"context"
	"errors"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
	"go.uber.org/mock/gomock"

	"github.com/gt-tech-ai/knowledge-engine/go/foundation/decorator"
	"github.com/gt-tech-ai/knowledge-engine/go/tests/fixtures"
	"github.com/gt-tech-ai/knowledge-engine/go/tests/mocks"
)

// foundFullStack composes all five generic decorators around base in the canonical
// §6.3 order (base → timeout → metrics → tracing → logging → recovery), the order
// the pipeline/workflow builders apply. onPanic is the tier's recovery contract.
func foundFullStack(
	base decorator.Executor[string, string],
	onPanic func(string, any) error,
) decorator.Executor[string, string] {
	e := base
	e = decorator.Timeout(e, time.Second)
	e = decorator.Metrics(
		e,
		"test",
		"svc",
		fixtures.NopMetrics(),
		decorator.DefaultBuckets,
	)
	e = decorator.Tracing(e, fixtures.NopTracer(), "test", "svc")
	e = decorator.Logging(e, fixtures.NopLogger(), "test", "svc")
	e = decorator.Recovery(e, fixtures.NopLogger(), "test", "svc", onPanic)
	return e
}

// TestDecoratorUnwrapSeam tests that every generic decorator implements the
// Unwrapper seam, so a test can walk the composed stack down to the base handler.
//
// Why this test is important:
//   - The Unwrap seam (mothership-parity audit F7) is the sanctioned way a test
//     reaches the underlying handler through the decorator stack; without it, a
//     test would have to reach into unexported decorator internals. It also proves
//     the wrap depth/order: five decorators sit between the top and the base.
//
// What it tests:
//   - Walking Unwrap from the composed top makes exactly five hops and terminates
//     at the exact base instance.
func TestDecoratorUnwrapSeam(t *testing.T) {
	t.Parallel()
	base := fixtures.PassPipeline()
	top := foundFullStack(base, func(string, any) error { return nil })

	hops := 0
	cur := decorator.Executor[string, string](top)
	for {
		u, ok := cur.(decorator.Unwrapper[string, string])
		if !ok {
			break
		}
		cur = u.Unwrap()
		hops++
	}

	require.Equal(t, 5, hops, "five decorators wrap the base")
	require.Same(
		t,
		base,
		cur.(*mocks.MockPipeline[string, string]),
		"Unwrap terminates at the exact base handler",
	)
}

// TestDecoratorRecoveryUsesOnPanic tests that the recovery decorator turns a panic
// into the tier-supplied error contract rather than a hardcoded one.
//
// Why this test is important:
//   - The pipeline tier returns a coded errors.Internal and the workflow tier a
//     plain wrap; the generic recovery must honor whichever onPanic the tier
//     supplies, or one tier's error contract would silently change.
//
// What it tests:
//   - A panicking base wrapped by Recovery returns exactly the onPanic error.
func TestDecoratorRecoveryUsesOnPanic(t *testing.T) {
	t.Parallel()
	sentinel := errors.New("recovered by contract")
	base := fixtures.PanicPipeline()

	rec := decorator.Recovery[string, string](
		base, fixtures.NopLogger(), "test", "svc",
		func(string, any) error { return sentinel },
	)

	_, err := rec.Execute(context.Background(), "in")
	require.ErrorIs(t, err, sentinel)
}

// TestDecoratorStackPassesResultThrough tests that a happy-path execution returns
// the base's result unchanged through the full five-decorator stack.
//
// Why this test is important:
//   - The decorators are transparent on the success path; a decorator that dropped
//     or altered the result would corrupt every pipeline/workflow that composes it.
//
// What it tests:
//   - The full stack over a base returning "out-<in>" yields "out-x" with no error.
func TestDecoratorStackPassesResultThrough(t *testing.T) {
	t.Parallel()
	base := fixtures.StubPipeline(func(_ context.Context, in string) (string, error) {
		return "out-" + in, nil
	})
	top := foundFullStack(base, func(string, any) error { return nil })

	got, err := top.Execute(context.Background(), "x")
	require.NoError(t, err)
	require.Equal(t, "out-x", got)
}

// TestMetricsDecoratorLabelsByInstanceName tests that the metrics decorator labels
// every series by the per-instance name, not the constant tier.
//
// Why this test is important:
//   - The label VALUE is the only dimension that separates one pipeline's metrics
//     from another's. Passing the tier as the value would collapse every pipeline
//     into a single "pipeline"-labelled series, so "which pipeline is slow or
//     erroring" becomes unanswerable. A prior refactor regressed exactly this and
//     a no-op metrics mock could not catch it; this spy asserts the label value.
//
// What it tests:
//   - After one Execute through Metrics(tier="pipeline", name="create-workspace"),
//     every recorded Counter/Histogram label value is "create-workspace" (the
//     instance name), never "pipeline" (the tier).
func TestMetricsDecoratorLabelsByInstanceName(t *testing.T) {
	t.Parallel()
	ctrl := gomock.NewController(t)

	// The mock Counter/Histogram capture the label VALUES the decorator passes to
	// Inc/Observe into labels, so the test can assert each series is labelled by the
	// per-instance name rather than the constant tier. Both counters (executions +
	// errors) and the histogram share the one labels slice, mirroring the three
	// series the metrics decorator constructs.
	var labels []string
	counter := mocks.NewMockCounter(ctrl)
	counter.EXPECT().Inc(gomock.Any()).Do(func(v ...string) {
		labels = append(labels, v...)
	}).AnyTimes()
	hist := mocks.NewMockHistogram(ctrl)
	hist.EXPECT().Observe(gomock.Any(), gomock.Any()).Do(func(_ float64, v ...string) {
		labels = append(labels, v...)
	}).AnyTimes()
	metricsMock := mocks.NewMockMetrics(ctrl)
	metricsMock.EXPECT().
		Counter(gomock.Any(), gomock.Any(), gomock.Any()).
		Return(counter).
		AnyTimes()
	metricsMock.EXPECT().
		Histogram(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
		Return(hist).
		AnyTimes()

	var inner decorator.Executor[string, string] = fixtures.StubPipeline(
		func(_ context.Context, in string) (string, error) { return in, nil },
	)
	m := decorator.Metrics(
		inner,
		"pipeline", "create-workspace",
		metricsMock,
		decorator.DefaultBuckets,
	)

	_, err := m.Execute(context.Background(), "x")
	require.NoError(t, err)

	require.NotEmpty(t, labels, "the metrics decorator recorded at least one series")
	for _, v := range labels {
		require.Equal(
			t,
			"create-workspace",
			v,
			"series are labelled by the instance name, not the tier",
		)
	}
}

File: python/techai_webutils/tests/unit/foundation_decorator_test.py
"""Tests for the generic decorator bases and Unwrap seam (foundation/decorator.py)."""

from __future__ import annotations

import time
from typing import Any
from unittest.mock import MagicMock

import pytest

from techai_webutils.core.errors.errors import AppTimeoutError, InternalError
from techai_webutils.foundation.decorator import (
    AsyncRecoveryDecoratorBase,
    AsyncTimeoutDecoratorBase,
    RecoveryDecoratorBase,
    TimeoutDecoratorBase,
    Unwrappable,
)
from techai_webutils.pipelines.base import BasePipeline
from techai_webutils.pipelines.decorators import (
    LoggingPipelineDecorator,
    MetricsPipelineDecorator,
)


class _SyncInner:
    """Minimal sync execute() double: returns a value, raises, or sleeps."""

    def __init__(self, *, result: Any = "ok", exc: Exception | None = None, delay: float = 0.0) -> None:
        self._result, self._exc, self._delay = result, exc, delay

    def execute(self, _input: Any) -> Any:  # noqa: ANN401
        if self._delay:
            time.sleep(self._delay)
        if self._exc is not None:
            raise self._exc
        return self._result


class _AsyncInner:
    """Minimal async execute() double: returns a value, raises, or sleeps."""

    def __init__(self, *, result: Any = "ok", exc: Exception | None = None, delay: float = 0.0) -> None:
        self._result, self._exc, self._delay = result, exc, delay

    async def execute(self, _input: Any) -> Any:  # noqa: ANN401
        if self._delay:
            import asyncio

            await asyncio.sleep(self._delay)
        if self._exc is not None:
            raise self._exc
        return self._result


def test_decorator_unwrap_returns_inner() -> None:
    """A decorator's unwrap() returns the exact inner it wraps, and it is Unwrappable.

    Why this test is important:
      - The Unwrap seam is the sanctioned way a test
        reaches the wrapped handler through the stack instead of touching decorator
        internals; every tier that subclasses the generic bases inherits it.

    What it tests:
      - unwrap() is the exact base instance, and isinstance(decorator, Unwrappable).
    """
    base = BasePipeline(lambda x: x)
    decorated = LoggingPipelineDecorator(base, "seam", MagicMock())

    assert decorated.unwrap() is base
    assert isinstance(decorated, Unwrappable)


def test_decorator_unwrap_walks_the_stack() -> None:
    """Walking unwrap() from the outermost decorator reaches the base in order.

    Why this test is important:
      - Proves the seam composes: a multi-decorator stack can be traversed layer by
        layer down to the underlying handler, which is what makes the seam useful.

    What it tests:
      - outer.unwrap() is the inner decorator; a second unwrap() is the base.
    """
    base = BasePipeline(lambda x: x)
    inner = MetricsPipelineDecorator(base, "seam", MagicMock())
    outer = LoggingPipelineDecorator(inner, "seam", MagicMock())

    assert outer.unwrap() is inner
    assert outer.unwrap().unwrap() is base


def test_sync_metrics_records_duration() -> None:
    """The sync metrics decorator observes an execution-duration sample.

    Why this test is important:
      - Metrics is the only sync base reached through a public class
        (MetricsPipelineDecorator); a broken observe() would silently drop every
        pipeline latency sample.

    What it tests:
      - execute() returns the inner result and records one histogram observation
        under the decorator's name.
    """
    histogram = MagicMock()
    decorated = MetricsPipelineDecorator(BasePipeline(lambda x: x.upper()), "m", histogram)

    assert decorated.execute("hi") == "HI"
    histogram.observe.assert_called_once()
    assert histogram.observe.call_args.kwargs["name"] == "m"


def test_sync_timeout_raises_apptimeout() -> None:
    """The sync timeout base raises AppTimeoutError when the inner exceeds the deadline.

    Why this test is important:
      - Timeout is the tier's backpressure valve; a slow unit must surface as a
        typed AppTimeoutError (not hang), so callers can shed rather than block.

    What it tests:
      - A 10ms deadline over a 200ms inner raises AppTimeoutError.
    """
    decorated = TimeoutDecoratorBase(_SyncInner(delay=0.2), 0.01, noun="pipeline")
    with pytest.raises(AppTimeoutError):
        decorated.execute("x")


def test_sync_recovery_normalizes_exceptions() -> None:
    """The sync recovery base re-raises AppError and wraps everything else.

    Why this test is important:
      - Recovery is the outermost guard; it must preserve typed AppErrors (so
        their codes reach the transport) yet never leak a raw exception.

    What it tests:
      - An AppError passes through unchanged; a plain exception becomes InternalError.
    """
    app_err = AppTimeoutError("boom")
    assert_passthrough = RecoveryDecoratorBase(_SyncInner(exc=app_err))
    with pytest.raises(AppTimeoutError) as caught:
        assert_passthrough.execute("x")
    assert caught.value is app_err

    wrapped = RecoveryDecoratorBase(_SyncInner(exc=ValueError("raw")))
    with pytest.raises(InternalError):
        wrapped.execute("x")


@pytest.mark.asyncio
async def test_async_timeout_raises_apptimeout() -> None:
    """The async timeout base raises AppTimeoutError when the inner exceeds the deadline.

    Why this test is important:
      - The async path is the one the pipeline/workflow builders actually compose;
        a broken timeout there would let a slow coroutine hang the whole request.

    What it tests:
      - A 10ms deadline over a 200ms coroutine raises AppTimeoutError.
    """
    decorated = AsyncTimeoutDecoratorBase(_AsyncInner(delay=0.2), 0.01, noun="async pipeline")
    with pytest.raises(AppTimeoutError):
        await decorated.execute("x")


@pytest.mark.asyncio
async def test_async_recovery_normalizes_exceptions() -> None:
    """The async recovery base re-raises AppError and wraps everything else.

    Why this test is important:
      - Same guarantee as the sync recovery but on the async path the builders use.

    What it tests:
      - An AppError passes through; a plain exception becomes InternalError.
    """
    app_err = AppTimeoutError("boom")
    passthrough = AsyncRecoveryDecoratorBase(_AsyncInner(exc=app_err))
    with pytest.raises(AppTimeoutError) as caught:
        await passthrough.execute("x")
    assert caught.value is app_err

    wrapped = AsyncRecoveryDecoratorBase(_AsyncInner(exc=ValueError("raw")))
    with pytest.raises(InternalError):
        await wrapped.execute("x")
